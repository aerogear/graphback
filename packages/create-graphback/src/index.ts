#!/usr/bin/env node

import * as yargs from 'yargs';

import { init } from './init/components';

type Params = {
  name?: string;
  templateName?: string;
  templateUrl?: string;
  filter?: string;
};

const command = '$0 <name>';

export const desc = 'Create Graphback project from available templates';

// tslint:disable-next-line: typedef
export const builder = (args: yargs.Argv) => {
  args.positional('name', {
    describe: 'Project name',
    type: 'string'
  });
  args.option('templateName', {
    describe: 'Name of the predefined template',
    type: 'string'
  });
  args.option('templateUrl', {
    describe:
      'GitHub URL of the template. For example (https://github.com/wtrocki/graphql-serve-example#master)',
    type: 'string'
  });
  args.option('filter', {
    describe:
      'A filter option used to limit the number of displayed templates that only matches it. For example "postgres" will only display templates that have a Postgres backed',
    type: 'string'
  });
};

export async function handler({
  name,
  templateName,
  templateUrl,
  filter
}: Params) {
  await init(name, templateName, templateUrl, filter);
}

if (require.main === module) {
  // eslint-disable-next-line no-unused-expressions
  yargs
    .command(command, desc, builder, handler)
    .example(
      '$0 my-awesome-project',
      'creates the "my-awesome-project" in the current working directory'
    )
    .demandCommand(1)
    .strict()
    .recommendCommands()
    .help()
    .alias('h', 'help')
    .version()
    .alias('v', 'version').argv;
}
