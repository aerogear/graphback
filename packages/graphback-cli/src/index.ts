#!/usr/bin/env node

import * as yargs from 'yargs';

if (require.main === module) {
  // eslint-disable-next-line no-unused-expressions
  yargs
    .commandDir('commands')
    .demandCommand(1)
    .strict()
    .recommendCommands()
    .help()
    .alias('h', 'help')
    .version()
    .alias('v', 'version')
    .argv;
}
