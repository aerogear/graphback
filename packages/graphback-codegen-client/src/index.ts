export * from './ClientDocuments'
export * from './ClientCRUDPlugin'