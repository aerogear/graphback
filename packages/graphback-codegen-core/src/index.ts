 
// Top level API
export * from './api'
export * from './graphql'
