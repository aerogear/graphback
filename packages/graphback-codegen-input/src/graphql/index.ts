
export * from "./GraphQLContextCreator"
export * from "./graphqlUtils"