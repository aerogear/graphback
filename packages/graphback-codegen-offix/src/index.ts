export * from './OffixPlugin'
export * from './OffixRuntimeData'


//Required for plugins
export { OffixPlugin as Plugin } from './OffixPlugin'