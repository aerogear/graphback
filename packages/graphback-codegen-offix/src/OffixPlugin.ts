
import { GraphbackCoreMetadata, GraphbackPlugin, ModelDefinition, getInputTypeName} from '@graphback/core'
import { GraphQLSchema } from 'graphql';
import { SchemaComposer } from 'graphql-compose';

/**
 * Configuration for Schema generator CRUD plugin
 */
export interface OffixPluginConfig {

    /**
     * RelativePath for the output files created by generator
     */
    outputPath: string


    /*
     * RelativePath for the output files created by generator
     */
    // TODO separate annotation for delta queries should be used for more granular support
    generateDeltaQueries: boolean

    /**
     * Delta resolvers format
     */
    deltaResolverFormat: 'ts' | 'js' | 'graphql',

}

export const SCHEMA_CRUD_PLUGIN_NAME = "OffixPlugin";

/**
 * Ofix plugin
 * 
 * Plugin is enabled by """ @datasync """ annotation
 * It will add version field to the schema and also generate diffQueries
 */
export class OffixPlugin extends GraphbackPlugin {

    private pluginConfig: OffixPluginConfig;

    public constructor(pluginConfig?: OffixPluginConfig) {
        super()
        this.pluginConfig = Object.assign({ format: 'graphql', outputFileName: 'schema' }, pluginConfig);
        if (!pluginConfig.outputPath) {
            throw new Error("schema plugin requires outputPath parameter")
        }
    }

    public transformSchema(metadata: GraphbackCoreMetadata): GraphQLSchema {
        const schema = metadata.getSchema()
        const schemaComposer = new SchemaComposer(schema);
        const models = metadata.getModelDefinitions();
        if (models.length === 0) {
            this.logWarning("Provided schema has no models. Returning original schema without any changes.")

            return schema;
        }

        models.forEach((model: ModelDefinition) => {
            // TODO use `versioned` marker to check if we should add version
            const modifiedType = schemaComposer.getOTC(model.graphqlType.name);
            modifiedType.addFields({
                version: 'Int'
            });

            const inputType = schemaComposer.getITC(getInputTypeName(model.graphqlType.name))
            if (inputType) {
                inputType.addFields({
                    version: 'Int'
                });
            }

            // Diff queries
            if (this.pluginConfig.generateDeltaQueries) {
                const diffQuery = `${model.graphqlType.name}Delta`
                schemaComposer.Query.addFields({
                    [diffQuery]: `[${model.graphqlType.name}]!`
                })
                schemaComposer.Query.addFieldArgs(diffQuery, {
                    lastSync: 'String!'
                })
                // TODO generate delta resolvers
            }
        })

        return schemaComposer.buildSchema()
    }

    public createResources(metadata: GraphbackCoreMetadata): void {
        // Schema plugin is going to create schema for us
        // No work to be done
    }

    public getPluginName() {
        return SCHEMA_CRUD_PLUGIN_NAME;
    }
}
