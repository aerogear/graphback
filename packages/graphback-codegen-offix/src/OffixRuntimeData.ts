import { PgKnexDBDataProvider, NoDataError } from "@graphback/runtime"
import { GraphQLObjectType } from 'graphql';
import { getDatabaseArguments } from '@graphback/core';


export class OffixDataProvider extends PgKnexDBDataProvider {
    public constructor(baseType: GraphQLObjectType, db: any) {
        super(baseType, db);
    }

    public async update(data: any): Promise<any> {
        const { idField, data: updateData } = getDatabaseArguments(this.tableMap, data);
        //tslint:disable-next-line: await-promise
        const dbResult = await this.db.select().from(this.tableName).where(idField.name, '=', idField.value);
        if (dbResult && dbResult[0]) {
            if (data.version !== dbResult[0].version) {
                throw { conflictInfo: { serverState: dbResult, clientState: data } }
            }

            //tslint:disable-next-line: await-promise
            const updateResult = await this.db(this.tableName).update(updateData).where(idField.name, '=', idField.value);
            if (updateResult === 1) {
                return dbResult[0]
            }
        }
        throw new NoDataError(`Cannot update ${this.tableName}`);
    }
}