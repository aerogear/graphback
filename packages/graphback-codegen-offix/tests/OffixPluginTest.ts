import { readFileSync } from 'fs';
import test, { ExecutionContext } from 'ava';
import { buildSchema, printSchema } from 'graphql';
import { GraphbackCoreMetadata } from '@graphback/core';
import { OffixPlugin } from '../src/OffixPlugin';

const schemaText = readFileSync(`${__dirname}/mock.graphql`, 'utf8')

test('Test snapshot config gql', async (t: ExecutionContext) => {
  const defautConfig = {
    "create": true,
    "update": true,
    "findAll": true,
    "find": true,
    "delete": true,
    "subCreate": true,
    "subUpdate": true,
    "subDelete": true
  }

  const schemaGenerator = new OffixPlugin({ outputPath: './tmp', generateDeltaQueries: true })
  const metadata = new GraphbackCoreMetadata({
    crudMethods: defautConfig
  }, buildSchema(schemaText))
  const schema = schemaGenerator.transformSchema(metadata)
  t.snapshot(printSchema(schema));
});
