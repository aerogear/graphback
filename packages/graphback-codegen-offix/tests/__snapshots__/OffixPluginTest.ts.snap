// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test snapshot config gql 1`] = `
"\\"\\"\\" @model \\"\\"\\"
type Comment {
  id: ID!
  title: String!
  description: String!

  \\"\\"\\"@manyToOne(field: 'comments', key: 'noteId')\\"\\"\\"
  noteComment: Note!
  version: Int
}

type Mutation {
  likeNote(id: ID!): Note!
}

\\"\\"\\" @model \\"\\"\\"
type Note {
  id: ID!
  title: String!
  description: String!

  \\"\\"\\"@oneToMany(field: 'noteComment', key: 'noteId')\\"\\"\\"
  comments: [Comment]!
  version: Int
}

type Query {
  getLikedNotes(id: ID!, names: [String]!): Note!
  NoteDelta(lastSync: String!): [Note]!
  CommentDelta(lastSync: String!): [Comment]!
}
"
`;
